### PR TITLE
Correctly report Static Mode Violation

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: src/eei.cpp [hunk 1]
```diff
@@ -426,7 +426,9 @@ string toHex(evm_uint256be const& value) {
 
       HERA_DEBUG << "log " << hex << dataOffset << " " << length << " " << numberOfTopics << dec << "\n";
 
-      heraAssert(!(msg.flags & EVM_STATIC), "\"log\" attempted in static mode");
+      if (msg.flags & EVM_STATIC)
+        throw StaticModeViolation{};
```
Comment (axic) [inline src/eei.cpp]: I'm not sure how beneficial it would be, but I think it may make sense keeping the function name which cause the violation.

Comment (chfast) [reply src/eei.cpp]: I can add it to the exception, but there is no place where to use it (we should be moving from cout to proper tracing).

Comment (axic) [reply src/eei.cpp]: We have those `HERA_DEBUGGING` lines. At least add `TODO` or `FIXME` where the exception is declared as a reminder. Plus create an issue for it.

File: src/eei.h [hunk 2]
```diff
@@ -39,6 +39,11 @@ class OutOfGasException : public std::exception {
   const char* what() const noexcept override { return "Out of gas."; }
 };
 
+class StaticModeViolation : public std::exception {
+public:
+  const char* what() const noexcept override { return "Static mode violation"; }
```
Comment (axic) [inline src/eei.h]: Dot missing (if want to be consistent with OOG).

**Comment by chfast** [comment]
Review again.

File: src/hera.cpp [hunk 3]
```diff
@@ -341,6 +341,8 @@ evm_result evm_execute(
     ret.gas_left = result.gasLeft;
   } catch (OutOfGasException const&) {
     ret.status_code = EVM_OUT_OF_GAS;
+  } catch (StaticModeViolation const&) {
+    ret.status_code = EVM_STATIC_MODE_ERROR;
```
Comment (axic) [inline src/hera.cpp]: Actually, since you added the reason, it makes sense to put a `HERA_DEBUGGING` case here displaying the reason.

**Pull request merged**